### ISSUE TITLE
BUG: RegularGridInterpolator returns Nan at input points

### ISSUE THREAD

**Issue by ytbaum** [scipy/scipy#19277]
### Describe your issue.

In my project, I take the following actions:

1. initialize a 2D array, call it Z, by setting all values to np.nan
2. set some, but not all, of the elements to real values
3. provide Z as the input values for RegularGridInterpolator. The value locations are two other arrays X and Y as created by meshgrid.
4. try to compute the values of the function Z at some locations (x,y) by interpolating. (x,y) sometimes coincide with locations of input data, but not always.

The values returned from step 4 are sometimes Nan even when they are at locations where I set a real value as in step 2.

Conceptually, this seems to me to be incorrect. I would expect that when I ask RegularGridInterpolator to compute a value at a point where I provided a value, it should just return the value that I provided.

This image provides an example:
![interior Z values defined](https://github.com/scipy/scipy/assets/339276/566cb8cb-4f7b-456f-b5f2-3b11df680d32)

The dots are at the locations of X and Y. Red dots show locations where a real value is returned by interpolation, and blue dots show locations where Nan is returned. I would expect the blue dots along the left (X = -4, Y in [-4, 4]) and bottom (Y = -4, X in [-4, 4]) edges of the contour map to be red.

This does not happen if all elements of Z are set to real values, as shown here:
![all Z values defined](https://github.com/scipy/scipy/assets/339276/a6ba6212-7983-4c95-8604-ed37fff54111)

(in this case all dots are blue because they are all defined). 

### Reproducing Code Example

```python
import numpy as np
from scipy.interpolate import RegularGridInterpolator as rgi
import matplotlib.pyplot as plt

x_axis = np.arange(-5, 6)
y_axis = np.arange(-5, 6)
X, Y = np.meshgrid(x_axis, y_axis)

# define Z on all grid nodes
Z = np.full_like(X.astype('float64'), np.nan)
Z = X**2 + Y**2

# interpolate
interp = rgi((x_axis, y_axis), Z, bounds_error = False)
interp_z = interp(np.vstack([X.flat, Y.flat]).T)

# make figure
fig = plt.figure()
plt.contourf(X, Y, Z)
interp_z_defined = ~np.isnan(interp_z)
print(interp_z_defined.all())
s_interp_defined = plt.scatter(X.flat, Y.flat, c = interp_z_defined.flat, cmap = 'seismic')
plt.title("All values of Z defined before interpolation")
plt.xlabel("X")
plt.ylabel("Y")

# define Z only on interior grid nodes
Z = np.full_like(X.astype('float64'), np.nan)
Z[1:-1, 1:-1] = X[1:-1, 1:-1]**2 + Y[1:-1, 1:-1]**2

# interpolate
interp = rgi((x_axis, y_axis), Z, bounds_error = False)
interp_z = interp(np.vstack([X.flat, Y.flat]).T)

# make figure
fig = plt.figure()
plt.contourf(X, Y, Z)
interp_z_defined = ~np.isnan(interp_z)
print(interp_z_defined.all())
s_interp_defined = plt.scatter(X.flat, Y.flat, c = interp_z_defined.flat, cmap = 'seismic')
plt.title("Only interior values of Z defined before interpolation")
plt.xlabel("X")
plt.ylabel("Y")
```

### Error message

```shell
There is no error message.
```

### SciPy/NumPy/Python version and system information

```shell
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
  NOT AVAILABLE
openblas_clapack_info:
  NOT AVAILABLE
flame_info:
  NOT AVAILABLE
accelerate_info:
  NOT AVAILABLE
atlas_3_10_threads_info:
  NOT AVAILABLE
atlas_3_10_info:
  NOT AVAILABLE
atlas_threads_info:
  NOT AVAILABLE
atlas_info:
  NOT AVAILABLE
lapack_info:
    libraries = ['lapack', 'blas', 'lapack', 'blas']
    library_dirs = ['C:/Users/user/Anaconda3/envs/clay_dep_cont_2023_06_12\\Library\\lib']
    language = f77
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
  NOT AVAILABLE
atlas_3_10_blas_threads_info:
  NOT AVAILABLE
atlas_3_10_blas_info:
  NOT AVAILABLE
atlas_blas_threads_info:
  NOT AVAILABLE
atlas_blas_info:
  NOT AVAILABLE
blas_info:
    libraries = ['cblas', 'blas', 'cblas', 'blas', 'cblas', 'blas']
    library_dirs = ['C:/Users/user/Anaconda3/envs/clay_dep_cont_2023_06_12\\Library\\lib']
    include_dirs = ['C:/Users/user/Anaconda3/envs/clay_dep_cont_2023_06_12\\Library\\include']
    language = f77
    define_macros = [('HAVE_CBLAS', None)]
blas_opt_info:
    define_macros = [('NO_ATLAS_INFO', 1), ('HAVE_CBLAS', None)]
    libraries = ['cblas', 'blas', 'cblas', 'blas', 'cblas', 'blas']
    library_dirs = ['C:/Users/user/Anaconda3/envs/clay_dep_cont_2023_06_12\\Library\\lib']
    include_dirs = ['C:/Users/user/Anaconda3/envs/clay_dep_cont_2023_06_12\\Library\\include']
    language = f77
lapack_opt_info:
    libraries = ['lapack', 'blas', 'lapack', 'blas', 'cblas', 'blas', 'cblas', 'blas', 'cblas', 'blas']
    library_dirs = ['C:/Users/user/Anaconda3/envs/clay_dep_cont_2023_06_12\\Library\\lib']
    language = f77
    define_macros = [('NO_ATLAS_INFO', 1), ('HAVE_CBLAS', None)]
    include_dirs = ['C:/Users/user/Anaconda3/envs/clay_dep_cont_2023_06_12\\Library\\include']
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2
    not found = AVX512F,AVX512CD,AVX512_SKX,AVX512_CLX,AVX512_CNL,AVX512_ICL
```

**Comment by ytbaum** [comment]
Sorry, omitted the first line of the system info. Here it is:

1.9.3 1.23.4 sys.version_info(major=3, minor=10, micro=6, releaselevel='final', serial=0)

**Comment by ev-br** [comment]
TL;DR: behavior with nans is undefined in scipy.interpolate; nans are sticky and tend to taint anything they touch.

Basically, this is a duplicate of https://github.com/scipy/scipy/issues/11381 -- but the illustration is nicer and clearer, thank you for it!

I'm going to close this request as a duplicate wontfix, feel free to continue discussing though.
